core-services/src/main/java/org/zalando/nakadi/validation/ResourceValidationHelperService.java

```diff
@@ -0,0 +1,148 @@
+    private static final String START_END_LABEL_ERROR = "Label should start and end with [a-zA-Z0-9].";
+    private static final String KEY_NAME_CHARACTER_ERROR = "Key name should use only [a-zA-Z0-9] and `-`, '_', and `.`";
+    private static final String LABEL_CHARACTER_ERROR = "Label should use only [a-zA-Z0-9] and `-`, '_', and `.`";
+    private static final String LONGER_PREFIX_ERROR = "Key prefix cannot be more than 253 characters.";
+    private static final String INVALID_PREFIX_ERROR = "Key prefix must follow RFC 1123 DNS subdomain format. \n" +
+            "It must contain one or more segments separated by '.'(dot), and each segment must start and end with " +
+            "[a-z0-9], and may contain - (hyphen) and [a-z0-9] in between.";
+    private static final String PREFIX_SEPARATOR = "/";
+    private static final String LONG_ANNOTATION_ERROR = "Annotation cannot be more than " + MAX_ANNOTATION_LENGTH
+            + " characters long.";
+    private static final String LONG_LABEL_ERROR = "LABEL cannot be more than " + MAX_LABEL_LENGTH + "charecters long.";
+
+    public void checkAnnotations(final ResourceAnnotations annotations) throws InvalidResourceAnnotationException {
+        if (annotations != null) {
+            annotations.forEach(this::validateAnnotation);
+        }
+    }
+
+    public void checkLabels(final ResourceLabels labels) throws InvalidResourceLabelException {
+        if (labels != null) {
+            labels.forEach(this::validateLabels);
+        }
+    }
+
+    private void validateAnnotation(final String key, final String value) {
+        validateKey(key).ifPresent(error -> {
+            throw new InvalidResourceAnnotationException(String.format(ERROR_MESSAGE_TEMPLATE, key, value, error));
+        });
+        if (value != null && value.length() > MAX_ANNOTATION_LENGTH) {
+            throw new InvalidResourceAnnotationException(
+                    String.format(ERROR_MESSAGE_TEMPLATE, key, value, LONG_ANNOTATION_ERROR));
+        }
+    }
+
+    private void validateLabels(final String key, final String value) {
+        validateKey(key).ifPresent(error -> {
+            throw new InvalidResourceLabelException(error);
+        });
+        if (value == null || value.isEmpty()) {
+            return;
+        }
+        if (value.length() > MAX_LABEL_LENGTH) {
+            throw new InvalidResourceLabelException(LONG_LABEL_ERROR);
+        }
+        if (!ALNUM_START_END_PATTERN.matcher(value).matches()) {
+            throw new InvalidResourceLabelException(START_END_LABEL_ERROR);
+        }
+        if (!ALLOWED_CHARACTER_PATTERN.matcher(value).matches()) {
+            throw new InvalidResourceLabelException(LABEL_CHARACTER_ERROR);
+        }
+    }
+
+    /*
+     * Both Annotation and Label key share the same format.
+     */
+    private Optional<String> validateKey(final String key) {
```

Comment:
Why not throw a `KeyValidationException`?  Returning optional error seems not Java-idiomatic.

Reply:
:+1: